compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -2347,16 +2347,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 suggested_field_name.to_string(),
                 Applicability::MaybeIncorrect,
             );
+            return;
+        }
+
+        let struct_variant_def = def.non_enum_variant();
+        let field_names = self.available_field_names(struct_variant_def, access_span);
+        let name = field.name.as_str();
+
+        fn get_common_prefix_len(a: &str, b: &str) -> usize {
+            let common_bytes_count =
+                a.as_bytes().iter().zip(b.as_bytes().iter()).take_while(|(c, d)| c == d).count();
+
+            return a.floor_char_boundary(common_bytes_count);
+        }
+
+        // Calculate the field programmer most likely wanted to use.
+        let longest_prefix_len = field_names
+            .iter()
+            .map(Symbol::as_str)
+            .map(move |s| get_common_prefix_len(s, name))
+            .max()
+            .unwrap_or(0);
+
+        // Field name will have at least one character.
+        let last_letter = name.chars().last().unwrap();
+        // When the keyboard gets hit, first there will be a few keystrokes registered,
+        // then one button will be "chosen" as continuation.
+        let gibberish = name[longest_prefix_len..].trim_end_matches(last_letter);
+        let letters_asleep = name.len() - longest_prefix_len - gibberish.len();
+
+        use std::time::Duration;
+        const KEY_REPEAT_PER_SEC: u64 = 25;
+        const KEY_REPEAT_DELAY: Duration = Duration::from_millis(600);
```

Comment:
I'm not sure these assumptions are reasonable. My key repeat delay, for example, is closer to 200 milliseconds. If this remains hard-coded, I (and I assume many other users) could miss out on the benefits of this diagnostic.

Could you instead get these values from the OS on platforms where there's a well-documented way to do that?

Otherwise, this looks fantastic. It's about time! :tada: :rocket: